content/en/blog/_posts/2021-06-08-Kubeadm-Avoid-Default-Address.md

```diff
@@ -0,0 +1,285 @@
+
+```yaml
+apiVersion: kubeadm.k8s.io/v1beta2
+kind: ClusterConfiguration
+apiServer:
+  extraArgs:
+    advertise-address: 192.168.56.101
+```
+
+## The Subject Alternative Name list in the kube-apiserver's certificate
+
+This job can be done with either a command line flag or a bit of
+config file contents.  The command line flag is
+`--apiserver-cert-extra-sans=192.168.56.101`.  The config file content
+sets the `apiServer.certSANS` of a `ClusterConfiguration` to hold the
+desired extra entries, such as `IP:192.168.56.101`.  Because the
+Subject Alternative Name list is a list, it is not functionally
+critical to make that list exclude the undesired default address.
+
+This job together with replacing the address
+that appears in the generated `kubeadm join` command can be done with
+a different command line flag or bit of configuration content.
+
+## The IP address in generated `kubeadm join` commands
+
+This is overridden with a command line flag or bit of config file
+contents.  This same bit of kubeadm configuration also gets the
+desired address into the kube-apisever's x509 certificate's Subject
+Alternative Name list.
+
+The command line flag is `--control-plane-endpoint=192.168.56.101`.
+
+The config file contents sets the `controlPlaneEndpoint` of a
+`ClusterConfiguration`.  Following is the cumulative config file
+content so far.
+
+```yaml
+apiVersion: kubeadm.k8s.io/v1beta2
+kind: ClusterConfiguration
+apiServer:
+  extraArgs:
+    advertise-address: 192.168.56.101
+controlPlaneEndpoint: 192.168.56.101
+```
+
+## Each node's InternalIP type of address
+
+The `kubeadm` tool takes no responsibility for configuring this.
+However, it is part of the problem at hand.  I solved it by looking
+into the `systemd` configuration for the kubelet and discovering that:
+- the kubelet command line includes `$KUBELET_EXTRA_ARGS`,
+- in a vanilla installation, nothing sets that variable,
+- the file `/etc/default/kubelet`, if it exists, is `source`d as part
+  of preparing the kubelet command line,
+- in a vanilla installation, that file does not exist.
+
+So I was able to use a simple solution that does not conflict with any
+maintained configuration.  On each node I create the file named
+`/etc/default/kubelet` and give it content like the following.
+
+```shell
+KUBELET_EXTRA_ARGS=--node-ip=192.168.56.101
+```
+
+Note: the kubelet also accepts a command line flag named `--address`,
+but that one does not get this job done.
+
+**Note well:** the proper content for this file is different on every
+node.
+
+## The grand finale
+
+With the above in place, I started creating my two-node cluster by
+executing
+
+```shell
+kubeadm init --config ka-init.conf
+```
+
+on the one control node and noting the `kubeadm join` command that was
+printed at the end.
+
+Next I used `kubectl` to look around an verify that everything looks
+good.  The kube-apiserver's Endpoints object looks good.
+
+```shell
+root@init1:~# kubectl get Endpoints -A
```

Comment:
Optionally: use `kubectl get endpointslice --all-namespaces` instead. Still works OK; bit more modern, long options are self-documenting.

Reply:
Yes, long options are better in a blog post.
Also, the display format for EndpointSlice objects is more efficient;
`kubectl get Endpoints` produces truncated output, but not so when querying EndpointSlice objects.